lib/Index/IndexRecord.cpp

```diff
@@ -393,6 +396,195 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
+  if (!M->getASTContext().LangOpts.EnableCXXInterop)
+    return;
+  // Skip system modules without an explicit 'cplusplus' requirement.
+  bool isSystem = clangModUnit->isSystemModule();
+  if (isSystem && !importer::requiresCPlusPlus(clangModule))
+    return;
+
+  // Make sure the `interfaces` directory is created.
+  std::string error;
+  if (initSymbolicInterfaceStorePath(indexStorePath, error)) {
+    diags.diagnose(SourceLoc(), diag::error_create_index_dir, error);
+    return;
+  }
+
+  // Determine the output name for the symbolic interface file.
+  clang::serialization::ModuleFile *ModFile =
+      clangCI.getASTReader()->getModuleManager().lookup(
+          clangModule->getASTFile());
+  assert(ModFile && "no module file loaded for module ?");
+  SmallString<128> interfaceOutputPath = indexStorePath;
+  appendSymbolicInterfaceToIndexStorePath(interfaceOutputPath);
+  appendSymbolicInterfaceClangModuleFilename(ModFile->FileName,
+                                             interfaceOutputPath);
+
+  // Check if the symbolic interface file is already up to date.
+  auto upToDate = isFileUpToDateForOutputFile(
+      interfaceOutputPath, StringRef(ModFile->FileName), error);
+  if (!upToDate) {
+    diags.diagnose(SourceLoc(), diag::error_index_failed_status_check, error);
+    return;
+  }
+  if (M->getASTContext().LangOpts.EnableIndexingSystemModuleRemarks) {
+    diags.diagnose(SourceLoc(), diag::remark_emitting_symbolic_interface_module,
+                   interfaceOutputPath, *upToDate);
+  }
+  if (*upToDate)
+    return;
+
+  // Output the interface to a temporary file first.
+  SmallString<128> tempOutputPath;
+  tempOutputPath = llvm::sys::path::parent_path(interfaceOutputPath);
+  llvm::sys::path::append(tempOutputPath,
+                          llvm::sys::path::filename(interfaceOutputPath));
+  tempOutputPath += "-%%%%%%%%";
+  int tempFD;
+  if (llvm::sys::fs::createUniqueFile(tempOutputPath.str(), tempFD,
+                                      tempOutputPath)) {
+    llvm::raw_string_ostream errOS(error);
+    errOS << "failed to create temporary file: " << tempOutputPath;
+    diags.diagnose(SourceLoc(), diag::error_write_index_record, errOS.str());
+    return;
+  }
+
+  llvm::raw_fd_ostream os(tempFD, /*shouldClose=*/true);
+  std::unique_ptr<ASTPrinter> printer;
+  printer.reset(new StreamPrinter(os));
+  ide::printSymbolicSwiftClangModuleInterface(M, *printer, clangModule);
+  os.close();
+
+  if (os.has_error()) {
+    llvm::raw_string_ostream errOS(error);
+    errOS << "failed to write '" << tempOutputPath
+          << "': " << os.error().message();
+    diags.diagnose(SourceLoc(), diag::error_write_index_record, errOS.str());
+    os.clear_error();
+    return;
+  }
+
+  // Move the resulting output to the destination symbolic interface file.
+  std::error_code ec = llvm::sys::fs::rename(
+      /*from=*/tempOutputPath.c_str(), /*to=*/interfaceOutputPath.c_str());
```

Comment:
I assume `rename` takes a `Twine` like the rest of the LLVM APIs, isn't there a constructor that takes a `SmallVectorImpl`? ie. I would have thought the `.c_str()`'s here weren't needed.

Reply:
good catch.